src/amber/router/pipe/flash.cr

```diff
@@ -29,17 +29,69 @@ module Amber
       end
 
       private def decode(flash, data)
-        json = Base64.decode_string(data)
-        values = JSON.parse(json)
+        yaml = Base64.decode_string(data)
+        values = YAML.parse(yaml)
         values.each do |key, value|
           flash[key.to_s] = value.to_s
         end
       end
 
       private def encode(flash)
-        data = Base64.encode(flash.to_json)
+        data = Base64.encode(flash.to_yaml)
         return data
       end
     end
   end
+
+  module Router
+    module Flash
+      # clear the flash messages.
+      def clear_flash
+        @flash = Hash.new
+      end
+
+      # Holds a hash of flash variables.  This can be used to hold data between
+      # requests. Once a flash message is read, it is marked for removal.
+      def flash
+        @flash ||= Hash.new
+      end
+
+      # A hash that keeps track if its been accessed
+      class Hash < Hash(String, String)
+        def initialize
+          @read = [] of String
+          super
+        end
+
+        def read 
+          @read
+        end
+
+        def []=(key : String | Symbol, value : V)
+          super(key.to_s, value)
+        end
+
+        def [](key : Symbol | String)
+          key = key.to_s
+          @read << key
+          fetch(key, nil)
+        end
+
+        def each
+          current = @first
+          while current
+            yield({current.key, current.value})
+            @read << current.key
+            current = current.fore
+          end
+          self
+        end
+
+        def unread
+          #TODO: unread marks them as read. Maybe fix this. It shouldn't actually matter since it get's reloaded next request.
+          reject { |key, _| @read.includes? key }
```

Comment:
Can be rewritten as `reject @read`, it'll delete every keys contained in `@read`